modules/deployment/src/main/java/org/mule/runtime/module/deployment/internal/DeploymentExecutor.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2023 Salesforce, Inc. All rights reserved.
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.module.deployment.internal;
+
+import static java.lang.System.getProperties;
+import static org.mule.runtime.api.util.MuleSystemProperties.SINGLE_DEPLOYMENT_PROPERTY;
+
+import org.mule.runtime.api.exception.MuleRuntimeException;
+import org.mule.runtime.module.artifact.api.Artifact;
+import org.mule.runtime.module.artifact.api.descriptor.DeployableArtifactDescriptor;
+import org.mule.runtime.module.deployment.api.DeploymentService;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.util.Optional;
+import java.util.Properties;
+
+import org.apache.commons.io.FileUtils;
+
+public class DeploymentExecutor {
+
+  private boolean artifactDeployed = false;
+  private final DeploymentService deploymentService;
+
+  public DeploymentExecutor(DeploymentService deploymentService) {
+    this.deploymentService = deploymentService;
+  }
+
+  public <D extends DeployableArtifactDescriptor, T extends Artifact<D>> void deploy(final URI artifactArchiveUri,
+                                                                                     final Optional<Properties> deploymentProperties,
+                                                                                     File artifactDeploymentFolder,
+                                                                                     ArchiveDeployer<D, T> archiveDeployer)
+      throws IOException {
+    if (!isSingleDeployment() || (isSingleDeployment() && !this.artifactDeployed)) {
+      this.artifactDeployed = true;
```

Comment:
verify that this is set as deployed after you know that you really deployed the artifact, not when you attempt to deploy it.